pyro/distributions/empirical.py

```diff
@@ -0,0 +1,160 @@
+        :param list buffer: list of new values.
+        :return: tensor with new values appended at the bottom.
+        """
+        buffer_tensor = torch.stack(buffer, dim=0)
+        return torch.cat([tensor, buffer_tensor], dim=0)
+
+    def _finalize(self):
+        """
+        Appends values collected in the samples/weights buffers to their
+        corresponding tensors.
+        """
+        if not self._samples_buffer:
+            return
+        self._samples = self._append_from_buffer(self._samples, self._samples_buffer)
+        self._log_weights = self._append_from_buffer(self._log_weights, self._weights_buffer)
+        self._categorical = Categorical(logits=self._log_weights)
+        # Reset buffers.
+        self._samples_buffer, self._weights_buffer = [], []
+
+    @property
+    def sample_size(self):
+        """
+        Number of samples that constitute the empirical distribution.
+
+        :return int: number of samples collected.
+        """
+        self._finalize()
+        if self._samples is None:
+            return 0
+        return self._samples.size(0)
+
+    def add(self, value, log_weight=None, weight=None):
+        """
+        Adds a new data point to the sample. The values in successive calls to
+        ``add`` must have the same tensor shape and size. Optionally, an
+        importance weight can be specified via ``log_weight`` or ``weight``
+        (default value of `1` is used if not specified).
+
+        :param torch.Tensor value: tensor to add to the sample.
+        :param torch.Tensor weight: log weight (optional) corresponding
+            to the sample.
+        :param torch.Tensor log_weight: weight (optional) corresponding
+            to the sample.
+        """
+        if is_validation_enabled():
+            if weight is not None and log_weight is not None:
+                raise ValueError("Only one of ```weight`` or ``log_weight`` should be specified.")
+            if torch.is_tensor(weight) and weight.dim() > 0:
+                raise ValueError("``weight.dim() > 0``, but weight should be a scalar.")
+
+        weight_type = torch.Tensor if value.dtype in (torch.int32, torch.int64) else value.type()
+        # Apply default weight of 1.0.
+        if log_weight is None and weight is None:
+            log_weight = torch.tensor(1.0).type(weight_type).log()
+        elif weight is not None and log_weight is None:
+            log_weight = math.log(weight)
+        if isinstance(log_weight, numbers.Number):
+            log_weight = torch.tensor(log_weight).type(weight_type)
+
+        # Seed the container tensors with the correct tensor types
+        if self._samples is None:
+            self._samples = value.new()
+            self._log_weights = log_weight.new()
+        # Append to the buffer list
+        self._samples_buffer.append(value)
+        self._weights_buffer.append(log_weight)
+
+    def sample(self, sample_shape=torch.Size()):
```

Comment:
(low priority comment:)

Is it sensible to define `rsample()`? If samples are differentiable, then the results should be differentiable wrt samples. The result is not differentiable wrt `weights` though. @martinjankowiak wdyt?